### PR TITLE
Refactoring change_flag

### PR DESCRIPTION
**Pull request by hennevogel**
* Don't do it via API anymore to get rid of ActiveXML
* Make flag operations resemble CRUD
* Don't use a crazy complicated helper/hand crafted javascript but a partial and unobtrusive javascript

### CODE REVIEW

File: src/api/app/mixins/webui/has_flags.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
+module Webui::HasFlags
+  def create_flag
+    authorize main_object, :update?
+
+    @flag = main_object.flags.new( status: params[:status], flag: params[:flag] )
+    @flag.architecture = Architecture.find_by_name(params[:architecture])
+    @flag.repo = params[:repository] unless params[:repository].blank?
+    @blah = 'whatever'
```
Comment (ChrisBr) [inline src/api/app/mixins/webui/has_flags.rb]: Huh?

Comment (hennevogel) [inline src/api/app/mixins/webui/has_flags.rb]: Ups

File: src/api/app/mixins/webui/has_flags.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    respond_to do |format|
+      if @flag.save
+        format.js {}
```
Comment (ChrisBr) [inline src/api/app/mixins/webui/has_flags.rb]: You don't need the parentheses!

File: src/api/app/mixins/webui/has_flags.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        format.json { render json: @flag.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def toggle_flag
+    authorize main_object, :update?
+
+    @flag = Flag.find(params[:flag])
+    if @flag.status == 'enable'
```
Comment (ChrisBr) [inline src/api/app/mixins/webui/has_flags.rb]: In one line:
``
@flag.status = @flag.status == 'enable' ? 'disable' : 'enable'
``

or if you introduce a ``enabled?`` / ``disabled?`` method:

`` @flag.status = @flag.enabled? ? 'disable' : 'enable' ``

But I'd move the whole to the flag model in a ``toogle`` (or ``toogle!``) function!

File: src/api/app/mixins/webui/has_flags.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      @flag.status = 'disable'
+    else
+      @flag.status = 'enable'
+    end
+
+    respond_to do |format|
+      if @flag.save
+        format.js {}
```
Comment (ChrisBr) [inline src/api/app/mixins/webui/has_flags.rb]: same here: you don't need the {}

File: src/api/app/mixins/webui/has_flags.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  def remove_flag
+    authorize main_object, :update?
+
+    @flag = Flag.find(params[:flag])
+    @project.flags.destroy(@flag)
+
+    respond_to do |format|
+      format.js {}
```
Comment (ChrisBr) [inline src/api/app/mixins/webui/has_flags.rb]: same here

File: src/api/app/models/flag.rb [hunk 6]
```diff
@@ -4,6 +4,28 @@ class Flag < ActiveRecord::Base
 
   belongs_to :architecture
 
+  validates :flag, :presence => true
+  validates :position, :presence => true
+  validates_numericality_of :position, :only_integer => true
+
+  before_validation(:on => :create) do
+    if self.project
+      self.position = (self.project.flags.maximum(:position) || 0 ) + 1
```
Comment (ChrisBr) [inline src/api/app/models/flag.rb]: Hm, we can't use here increment!, can we?
But then we have to make this thread safe (e.g. wrap in a transaction)

File: src/api/app/models/flag.rb [hunk 7]
```diff
@@ -4,6 +4,28 @@ class Flag < ActiveRecord::Base
...
+    elsif self.package
+      self.position = (self.package.flags.maximum(:position) || 0 ) + 1
```
Comment (ChrisBr) [inline src/api/app/models/flag.rb]: Same here!

File: src/api/app/models/flag.rb [hunk 8]
```diff
@@ -4,6 +4,28 @@ class Flag < ActiveRecord::Base
...
+    end
+  end
+
+  validate :validate_custom_save
+  def validate_custom_save
+    errors.add(:name, 'Please set either project or package.') if self.project.nil? and self.package.nil?
```
Comment (ChrisBr) [inline src/api/app/models/flag.rb]: ``
if self.project.nil? && self.package.nil?
``

File: src/api/app/models/flag.rb [hunk 9]
```diff
@@ -4,6 +4,28 @@ class Flag < ActiveRecord::Base
...
+    errors.add(:name, 'Please set either project or package.') unless self.project.nil? or self.package.nil?
```
Comment (ChrisBr) [inline src/api/app/models/flag.rb]: `` unless self.project.nil? || self.package.nil? ``

File: src/api/app/models/flag.rb [hunk 10]
```diff
@@ -4,6 +4,28 @@ class Flag < ActiveRecord::Base
...
+    errors.add(:flag, 'There needs to be a valid flag.') unless FlagHelper::TYPES.has_key?(self.flag.to_s)
+    # rubocop:disable Metrics/LineLength
+    errors.add(:status, 'Status needs to be enable or disable') unless (self.status && (self.status.to_sym == :enable or self.status.to_sym == :disable))
```
Comment (ChrisBr) [inline src/api/app/models/flag.rb]: ``
unless (self.status && (self.status.to_sym == :enable || self.status.to_sym == :disable))
``

File: src/api/app/models/flag.rb [hunk 11]
```diff
@@ -67,26 +93,13 @@ def to_s
...
-
-  before_validation(:on => :create) do
-    if self.project
-      self.position = (self.project.flags.maximum(:position) || 0 ) + 1
-    elsif self.package
-      self.position = (self.package.flags.maximum(:position) || 0 ) + 1
-    end
+  def name
+    name = flag
+    name += "-#{repo}" unless repo.blank?
```
Comment (ChrisBr) [inline src/api/app/models/flag.rb]: name << "-#{repo}" unless repo.blank?

File: src/api/app/models/flag.rb [hunk 12]
```diff
@@ -67,26 +93,13 @@ def to_s
...
+    name += "-#{architecture.name}" unless architecture_id.blank?
```
Comment (ChrisBr) [inline src/api/app/models/flag.rb]: ``
name << "-#{architecture.name}" unless architecture_id.blank?
``

**Comment by ChrisBr** [comment]
What about tests?

Otherwise LGTM

File: src/api/app/models/flag.rb (hunk 6 above)
Comment (hennevogel) [inline src/api/app/models/flag.rb]: I just moved this to the top of the file. Some nasty abuse to refactor for another day... :-)

File: src/api/app/models/flag.rb (hunk 8 above)
Comment (hennevogel) [inline src/api/app/models/flag.rb]: same here

File: src/api/app/models/flag.rb (hunk 9 above)
Comment (hennevogel) [inline src/api/app/models/flag.rb]: same here, just moved

File: src/api/app/models/flag.rb (hunk 10 above)
Comment (hennevogel) [inline src/api/app/models/flag.rb]: just moved

File: src/api/app/models/flag.rb (hunk 11 above)
Comment (hennevogel) [inline src/api/app/models/flag.rb]: how is that different?

Comment (coolo) [inline src/api/app/models/flag.rb]: name += "a" is actually
name = name.concat("a") - so it creates an iterim object which is then assigned

while name << "a" is just name.concat("a") - no extra assignment

Comment (coolo) [inline src/api/app/models/flag.rb]: (just saying)

Comment (hennevogel) [inline src/api/app/models/flag.rb]: So there is no difference :-)

Comment (coolo) [inline src/api/app/models/flag.rb]: it won't be a noticable performance gain, but it's a good habit to write performant code when having a choice ;)

**Comment by hennevogel** [comment]
Test are not that simple...

**Comment by hennevogel** [comment]
The current interface is untestable with capybara/poltergeist because of the way the links appear one hover. I could only remove the remove_flag (taking the default) option from the interface and make it a simple toggle...

**Comment by hennevogel** [comment]
A new mixin and page.execute_script in test. I'll better go to the weekend and never come back... :disappointed:

**Comment by coolo** [comment]
volkswagen gem? :)

**Comment by hennevogel** [comment]
I give up. Wrote some tests, skipped them and added a FIXME.

**Comment by hennevogel** [comment]
So @ChrisBr can we merge this now? :-)

File: src/api/app/models/flag.rb (hunk 6 above)
Comment (ChrisBr) [inline src/api/app/models/flag.rb]: Ok!

**Comment by ChrisBr** [comment]
If you rebase ;)

**Pull request merged**